src/lib/comms/sol-coap.c

```diff
@@ -1018,22 +1037,50 @@ sol_coap_notify(struct sol_coap_server *server,
             sol_network_link_addr_to_str(&o->cliaddr, &addr);
             SOL_WRN("Failed to enqueue packet %p to %.*s", header,
                 SOL_STR_SLICE_PRINT(sol_buffer_get_slice(&addr)));
-            goto error_pkt;
+            goto err_header;
         }
 
+        if (cb)
+            sol_coap_packet_unref(pkt);
         sol_coap_packet_unref(header);
     }
 
-    sol_coap_packet_unref(pkt);
+    if (!cb)
+        sol_coap_packet_unref(pkt);
```

Comment:
You messing the variable pkt around, using for different purposes. I suggest that you stop trying to unref it here. unref it in sol_coap_notify.